design/20221205-memory-management.md

```diff
@@ -0,0 +1,500 @@
+
+##### Pros
+
+- In most setups in majority of cases where a control loop needs a `Secret` it would still be retrieved from cache (as it is certificate secrets that get parsed most frequently and those will be labelled in practically all cases)
+
+- Once graduated to GA would work for all installations without needing to discover a flag to set
+
+- data of unlabelled `Secret`s is not retrieved from kube apiserver at all
+
+#### C: Filter the Secrets to watch with a label
+
+This proposal suggests only watching `Secret`s with known `cert-manager.io` labels and annotations.
+The majority of `Secret`s for `Certificate`s will already have these applied by cert-manager (see [Secrets for Certificates](#secrets-for-certificates)). I cannot think of any cases where there would be a need to observe an event for a certificate secret without a label or annotation.
+Some of the [`Secret`s related to issuers](#secrets-for-clusterissuers) would need to be labelled manually by users. We can help them to discover which `Secret`s that are currently deployed to cluster and need labelling with a `cmctl` command.
+
+##### Cons
+
+- Bad user experience - breaking change to adopt and introduces a potential footgun after adoption as even if users labelled all relevant `Secret`s in cluster at time of adoption, there would likely be no visible warning if an unlabelled `Secret` for an issuer got created at some point in future and things would just silently not work (i.e `Secret` data updates would not trigger issuer reconcile etc).
+
+- This feature would likely need to be opt-in 'forever' as else it would be a major breaking change when adopting and a potential footgun after adoption
+
+- Future changes to cert-manager might need to take this into account (i.e if a new user created `Secret` is created for an issuer the cmctl command would also need to be updated, which could be easily forgotten)
+
+##### Pros
+
+- In terms of resource consumption and calls to apiserver, this would be the best solution (only relevant secrets watched and always retrieved from cache)
+
+#### D: Filter which Secrets to _not_ watch by a label
+
+This proposal suggests filtering w
+
+##### Pros
+
+- Could be pushed straight to GA
+
+- Straightforward to adopt, no footguns, does not complicate issuance model, no additional maintenance cost (future changes to cert-manager, such as adding a new control loop don't have to think about this)
+
+##### Cons
+
+- Perhaps not so good user experience- user needs to discover that there is a flag that needs to be set in order for cert-manager to not watch unrelated `Secret`s 
+
+##### Assumptions
+
+- the excessive memory consumption comes from known labelled `Secret`s (i.e helm release) rather than a large number of random `Secret`s in cluster.
+Note: would be good to get some user input on this.
+
+See an implementation for cainjector in https://github.com/cert-manager/cert-manager/pull/5174 thanks to @aubm for working on this.
+
+#### Comparison
```

Comment:
It would be nice to add another parameter to this comparison: Average certificate issuance time. It will help us to quantify the increase in certificate issuance time(if any).

Another bonus point would be to plot the same metric for different issuer types.

Reply:
Thanks @sathyanarays that's a good idea. I've added a rough comparison of issuance times https://github.com/irbekrm/cert-manager/blob/2ddd8c2413a26a6108093ba24c42615fcbd11476/design/20221205-memory-management.md#issuance-of-a-large-number-of-certificates.

What will make the difference will be whether the users have labelled any `Secret`s referenced by the issuer, the issuer type should not really matter.

It'd be interesting to hear from some users with larger deployments/shorter lived certs how big an increase in issuance time is 'too big'

Reply:
I had a look at the script used. Polling is being done to check if all the certificates are issued: https://gist.github.com/irbekrm/bc56a917a164b1a3a097bda483def0b8#file-measure-issuance-time-sh-L76-L98. Will that not affect the performance parameters? can we confirm this by replacing `sleep 1` with `sleep 60`?

Reply:
> Is this being handled by the PR https://github.com/cert-manager/cert-manager/pull/5660?

No, we will not apply labels to user created issuer Secrets

Reply:
> I had a look at the script used. Polling is being done to check if all the certificates are issued: https://gist.github.com/irbekrm/bc56a917a164b1a3a097bda483def0b8#file-measure-issuance-time-sh-L76-L98. Will that not affect the performance parameters? can we confirm this by replacing sleep 1 with sleep 60?

That's a good point, but in general I would like folks to not look at those measurements as the 'average issuance time' because that would hugely vary depending on other factors, I ran the tests on a local kind cluster, it might be different on managed/'busy' Kubernetes etc. The script should only be used to get a rough estimate of a diff when run against different cert-manager versions, so shouldn't matter whether the apiserver is slowed down a bit by some additional requests?